modules/core/src/main/scala/com.snowplowanalytics.iglu/client/validator/ValidatorF.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2012-2021 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.iglu.client
+package validator
+
+// cats
+import cats.Functor
+import cats.data.NonEmptyList
+import cats.implicits._
+
+// circe
+import io.circe.Json
+
+/**
+ * Validates that Json conforms to a schema
+ *
+ * Unlike a simple [[Validator]], the results returned by a ValidatorF are wrapped in an effect type `F`.
+ * Implementations of ValidatorF are expected to use caches for more efficient validation.
+ */
+trait ValidatorF[F[_], K, A] {
```

Comment:
Is this trait really needed? Is it something we would like to expose publicly? I know it's like the original `Validator` but maybe we could call new methods from `CirceValidator` directly?

Reply:
I don't have a strong reason why it should be a trait, except that I copied the original `Validator`.  I guess it exists as a trait so that you could build a `Client` around json4s or some other json library.  We have [historical support for json4s](https://github.com/snowplow/iglu-scala-core/tree/1.1.0/iglu-core-json4s/src).  On the other hand, we are now so deeply committed to circe that I don't see any call to support more than one json library.  So yeah I guess we could kill the trait and call CirceValidator methods directly.

Reply:
If we make everything private to client, make new `Client` the only public thing, we can do everything we want with internals in the future if needed, right?

Reply:
Tried this in https://github.com/snowplow/iglu-scala-client/pull/203/commits/597f025fcb60d3ac09d09efa458e7d03e96f41ef